src/StreamJsonRpc.Tests/DelimitedMessageHandlerTests.cs

```diff
@@ -111,6 +111,33 @@ public void WriteAsync_PreferOperationCanceledException_AtEntry()
         await Assert.ThrowsAnyAsync<OperationCanceledException>(() => writeTask);
     }
 
+    /// <summary>
+    /// Tests that by default, WriteCoreAsync calls cannot be called again till
+    /// the Task returned from the prior call has completed.
+    /// </summary>
+    /// <param name="optimized">A value indicating whether we simulate a derived type that expresses readiness for a perf optimization.</param>
+    [Theory]
+    [InlineData(false)]
+    [InlineData(true)]
+    public async Task WriteAsync_SemaphoreIncludesWriteCoreAsync_Task(bool optimized)
+    {
+        var handler = new DelayedWriter(this.sendingStream, this.receivingStream, Encoding.UTF8);
+        handler.WriteCoreAsyncCallsWriteBeforeReturningPublic = optimized;
+        var writeTask = handler.WriteAsync("content", CancellationToken.None);
+        var write2Task = handler.WriteAsync("content", CancellationToken.None);
+
+        if (!optimized)
+        {
+            // Give the library extra time to do the wrong thing asynchronously.
```

Comment:
The thing I had in mind was that it might invoke my `WriteCoreAsync` method twice when it's only expected to happen once. I want to give it time to have called it twice before claiming success that it only called it once.